### PR TITLE
Add Date, POSIXct and hms types

### PR DESCRIPTION
**Pull request by anderic1**
Following #319 this PR adds support for Date, datetime and hms types. It depends on BH package version 1.75.0-0 which is not yet on CRAN.
The new option `extended_types` is set to `TRUE` by default in the `dbConnect()` method but could be set to `FALSE` to return to the old behaviour.

### CODE REVIEW

**Comment by krlmlr** [comment]
Thanks for looking into it. Is there a timeline for BH 1.75?

Can we keep `extended_types = FALSE` for now?

**Comment by anderic1** [comment]
BH will be updated in the beginning of January. Some depending packages needed to be updated first.

I hoped extended_types would be set to true, but I understand your concerns. You can go ahead and change it, I will not have access to my computer for the next few weeks.

File: src/SqliteResultImpl.cpp [hunk 1]
```diff
@@ -164,7 +165,20 @@ List SqliteResultImpl::get_column_info() {
...
-    types[i] = Rf_type2char(DbColumnStorage::sexptype_from_datatype(types_[i]));
+    switch(types_[i]) {
+    case DT_DATE:
+      types[i] = "Date";
+      break;
+    case DT_DATETIME:
+      types[i] = "POSIXct";
+      break;
+    case DT_TIME:
+      types[i] = "hms";
```
Comment (krlmlr) [inline src/SqliteResultImpl.cpp:176]: This needs to be the equivalent of `c("hms", "difftime")` .

File: DESCRIPTION [hunk 2]
```diff
@@ -43,14 +43,15 @@ Imports:
     memoise,
     methods,
     pkgconfig,
-    Rcpp (>= 0.12.7)
+    Rcpp (>= 0.12.7),
+    hms
```
Comment (krlmlr) [inline DESCRIPTION:47]: I'm a bit hesitant to bring in hms as a dependency here. Can we move to Suggests and load only if a connection with extended types is requested?

File: src/SqliteColumnDataSource.cpp [hunk 3]
```diff
@@ -3,14 +3,26 @@
...
+  with_alt_types(with_alt_types_)
 {
 }
 
 DATA_TYPE SqliteColumnDataSource::get_data_type() const {
+
+  if (with_alt_types) {
+      DATA_TYPE decl_dt = get_decl_data_type();
+      if (decl_dt == DT_DATE || decl_dt == DT_DATETIME || decl_dt == DT_TIME) {
+          return decl_dt;
```
Comment (krlmlr) [inline src/SqliteColumnDataSource.cpp:22]: Does this mean that a declared data type overrides what's stored in the columns?

File: src/SqliteColumnDataSource.cpp [hunk 4]
```diff
@@ -80,13 +92,51 @@ SEXP SqliteColumnDataSource::fetch_blob() const {
...
+      try {
+          bg::date dt(bg::from_simple_string(dtstr));
+          bg::date_duration delta = dt - bg::date(1970, 1, 1);
+          dateval = static_cast<double>(delta.days());
+      } catch (...) {
+          dateval = NA_REAL;
+      }
+      return dateval;
+  } else {
+      return static_cast<double>(sqlite3_column_int(get_stmt(), get_j()));
```
Comment (krlmlr) [inline src/SqliteColumnDataSource.cpp:113]: What happens if the underlying data is an integer or a blob?

**Review by krlmlr — COMMENTED** [review]
Thanks, I took a closer look.

Is the implementation in line with the discussion in https://github.com/r-dbi/RSQLite/issues/237?

**Comment by krlmlr** [comment]
When done, we can enable many more DBItest tests that test for roundtrip for these data types.

**Comment by anderic1** [comment]
Replies inline

> Sent: Sunday, December 27, 2020 at 10:27 AM
> From: "Kirill Müller" <notifications@github.com>
> To: "r-dbi/RSQLite" <RSQLite@noreply.github.com>
> Cc: "anderic1" <anderic1@gmx.com>, "Author" <author@noreply.github.com>
> Subject: Re: [r-dbi/RSQLite] Add Date, POSIXct and hms types (#333)
>
> @krlmlr commented on this pull request.
> 
> Thanks, I took a closer look.
> 
> Is the implementation in line with the discussion in https://github.com/r-dbi/RSQLite/issues/237?

I would say so, insofar as if the flag is not set everything will be like before. If (1) flag is set and (2) column is declared exactly modulo case, as 'date': a number will be interpreted as count of days since 1970-01-01, with class attribute 'Date' attached to it, ie an R date, and an iso string ('yyyy-mm-dd') will be converted to an R date. Everything else will return NA. Similarly for 'datetime'/'timestamp' -> POSIXct (UTC), and 'time' -> hms/difftime (unit=seconds).

> 
> > @@ -164,7 +165,20 @@ List SqliteResultImpl::get_column_info() {
>  
>    CharacterVector types(cache.ncols_);
>    for (size_t i = 0; i < cache.ncols_; i++) {
> -    types[i] = Rf_type2char(DbColumnStorage::sexptype_from_datatype(types_[i]));
> +    switch(types_[i]) {
> +    case DT_DATE:
> +      types[i] = "Date";
> +      break;
> +    case DT_DATETIME:
> +      types[i] = "POSIXct";
> +      break;
> +    case DT_TIME:
> +      types[i] = "hms";
> 
> This needs to be the equivalent of `c("hms", "difftime")` .
> 
Yes I was unsure about this one. POSIXct should similarly be c(POSIXct, POSIXt). Then the column should be a list column(?), but this also makes it harder to use. That's why I opted for a single element. 
> > @@ -43,14 +43,15 @@ Imports:
>      memoise,
>      methods,
>      pkgconfig,
> -    Rcpp (>= 0.12.7)
> +    Rcpp (>= 0.12.7),
> +    hms
> 
> I'm a bit hesitant to bring in hms as a dependency here. Can we move to Suggests and load only if a connection with extended types is requested?

I brought it in because as the existing code was written, at one moment the c++ code calls new_hms() from the hms package on the column. This could of course be changed, then hms can be moved to suggests. 
> 
> >  {
>  }
>  
>  DATA_TYPE SqliteColumnDataSource::get_data_type() const {
> +
> +  if (with_alt_types) {
> +      DATA_TYPE decl_dt = get_decl_data_type();
> +      if (decl_dt == DT_DATE || decl_dt == DT_DATETIME || decl_dt == DT_TIME) {
> +          return decl_dt;
> 
> Does this mean that a declared data type overrides what's stored in the columns?

Yes, but again only for the aforementioned column types, and if the flag is set. If the value cannot be converted NA is returned. If the flag is not set everything works like before.
> 
> > +  int dt = get_column_type();
> +
> +  if (dt == SQLITE_TEXT) {
> +      const char* dtstr = reinterpret_cast<const char*>(sqlite3_column_text(get_stmt(), get_j()));
> +      double dateval;
> +
> +      try {
> +          bg::date dt(bg::from_simple_string(dtstr));
> +          bg::date_duration delta = dt - bg::date(1970, 1, 1);
> +          dateval = static_cast<double>(delta.days());
> +      } catch (...) {
> +          dateval = NA_REAL;
> +      }
> +      return dateval;
> +  } else {
> +      return static_cast<double>(sqlite3_column_int(get_stmt(), get_j()));
> 
> What happens if the underlying data is an integer or a blob?
> 
Integer works as expected. Blob however i have not tested. This is an oversight, my guess is that it will return 0 (ie 1970-01-01). A better option imo would be to return NA in this case.

**Comment by anderic1** [comment]
I addressed all but one of the issues you mentioned and added warnings when blobs are encountered. I did not yet address the columninfo for time columns. As I mentioned earlier, how should this be returned? Should the whole column be coerced into a list column? Or perhaps add a third column as you alluded to in the beginning? For simplicity I think keeping just `"hms"`for time columns and `"POSIXct"` for datetime columns is the easiest and conveys sufficient information, but you be the judge.

**Comment by krlmlr** [comment]
Thanks. Could you please rerun (e.g. merge with master) when BH 1.75 becomes available on CRAN?

The class needs to be `c("hms", "difftime")`, for the column info the first component should be sufficient -- don't need to convert to a list here.

**Comment by anderic1** [comment]
One last thing, would you mind if I changed the default value of extended_types to something like `extended_types = getOption("RSQLite_extended_types", FALSE)`? That way it can easily be enabled globally if one so wishes.

**Comment by anderic1** [comment]
BH has been updated on cran. 
Some checks are still failing, the Windows and Mac images have not yet been updated it seems.
The extended_types setting is set to FALSE by default, but can be overridden by setting options(RSQLite_extended_types = TRUE). Some changes were also needed to fix checks on all platforms.

File: src/DbColumnStorage.cpp [hunk 5]
```diff
@@ -206,9 +209,17 @@ SEXP DbColumnStorage::set_attribs_from_datatype(SEXP x, DATA_TYPE dt) {
   case DT_BLOB:
     return new_blob(x);
 
-  case DT_TIME:
-    return new_hms(x);
+  case DT_TIME: {
+      Rcpp::RObject ro = Rcpp::RObject(x);
```
Comment (krlmlr) [inline src/DbColumnStorage.cpp:213]: What's the motivation for this change?

**Review by krlmlr — COMMENTED** [review]
Thanks, looks good. Just a few minor tweaks necessary.

File: R/connect.R [hunk 6]
```diff
@@ -114,7 +131,8 @@ setMethod("dbConnect", "SQLiteDriver",
   function(drv, dbname = "", ..., loadable.extensions = TRUE,
            default.extensions = loadable.extensions, cache_size = NULL,
            synchronous = "off", flags = SQLITE_RWC, vfs = NULL,
-           bigint = c("integer64", "integer", "numeric", "character")) {
+           bigint = c("integer64", "integer", "numeric", "character"),
+           extended_types = getOption("RSQLite_extended_types", FALSE)) {
```
[the remark this answers is not in the document]
Comment (anderic1) [reply R/connect.R]: ok

File: R/connect.R [hunk 7]
```diff
@@ -128,14 +146,17 @@ setMethod("dbConnect", "SQLiteDriver",
 
     bigint <- match.arg(bigint)
 
+    extended_types <- isTRUE(extended_types)
+
```
[the remark this answers is not in the document]
Comment (anderic1) [reply R/connect.R:150]: My bad I misunderstood your earlier intention to remove hms from imports to suggests.
My idea behind it was allowing it even if hms is not installed, since in this case it will behave as a normal difftime object.
Could we perhaps put :
```R
if (extended_types) {
      if (!requireNamespace("hms", quietly = TRUE)) {
        message("The hms package is not installed.")
      }
    }
```
or similar, ie a simple message instead of erroring out. If hms is found time columns will print nicely, if not they will print as normal difftime objects (Time difference of xxx secs).
If you prefer otherwise, your suggested changes are fine.

File: src/DbColumnStorage.cpp (hunk 5 above)
Comment (anderic1) [reply src/DbColumnStorage.cpp:213]: Cf above since calling new_hms depends on the hms package

File: R/connect.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (krlmlr) [reply R/connect.R:150]: I'd rather fail cleanly and loudly here. The alternative would be a message and returning objects of class `"difftime"` (not `c("hms", "difftime")`) if hms is not installed); otherwise behavior changes if hms is attached later on. This is clearly too much effort.

[the remark this answers is not in the document]
Comment (krlmlr) [reply R/connect.R:150]: Also, it's easier to relax later than to tighten restrictions after the fact.

**Comment by anderic1** [comment]
I think we are done here. I committed your suggestions and you are right, it is better to be too strict and then loosen up than the other way around.

**Comment by krlmlr** [comment]
Thanks!
